File: source/com/barrybecker4/simulation/liquid/model/CellVelocity.java
/** Copyright by Barry G. Becker, 2000-2011. Licensed under MIT License: http://www.opensource.org/licenses/MIT  */
package com.barrybecker4.simulation.liquid.model;

/**
 *  Keeps track of the current and last velocities for the cell.
 *  The u component heads out of the cell to the left, and
 *  the v component heads out of the cell to the top.
 *  velocities in x, y directions are defined at the center of each face.
 *
 *  @author Barry Becker
 */
public class CellVelocity {

    /** uip_ = u(i+0.5, j, k) */
    private final double[] uip_ = new double[2];

    /** vjp_ = v(i, j+0.5, k) */
    private final double[] vjp_ = new double[2];

    /** use this to switch between current and last copies of fields. (hack) */
    private int current_;

    /**
     * Constructor
     */
    public CellVelocity()  {
        uip_[0] = uip_[1] = 0;
        vjp_[0] = vjp_[1] = 0;
    }

    /**
     *  global swap of fields (use with care). (hack)
     */
    public void step()  {
        current_ = 1 - current_;
    }

    public void passThrough() {
        uip_[1 - current_] = uip_[current_]; // + dt * forceX;
        vjp_[1 - current_] = vjp_[current_]; // + dt * forceY;
    }

    public void initializeU(double u) {
        uip_[0] = u;
        uip_[1] = u;
    }
    public void initializeV(double v) {
        vjp_[0] = v;
        vjp_[1] = v;
    }

    public double getU() {
        return uip_[current_];
    }
    public double getV() {
        return vjp_[current_];
    }

    public void setCurrentU(double u) {
        uip_[current_] = u;
    }

    public void setCurrentV(double v) {
        vjp_[current_] = v;
    }

    public void setNewU(double newu) {
        uip_[1 - current_] = newu;
    }

    public void setNewV(double newv) {
        vjp_[1 - current_] = newv;
    }

    public void incrementU(double inc) {
        uip_[current_] += inc;
    }

    public void incrementV(double inc) {
        vjp_[current_] += inc;
    }

    public void initialize(double u, double v) {
        uip_[0] = u;
        uip_[1] = u;
        vjp_[0] = v;
        vjp_[1] = v;
    }

    @Override
    public String toString() {
        StringBuilder sb = new StringBuilder("velocity=(");
        sb.append(uip_[current_]).append(", ").append(vjp_[current_]).append(")");
        return sb.toString();
    }
}

File: scala-source/com/barrybecker4/simulation/liquid/model/CellVelocity.scala
// Copyright by Barry G. Becker, 2016-2017. Licensed under MIT License: http://www.opensource.org/licenses/MIT
package com.barrybecker4.simulation.liquid.model

/**
  * Keeps track of the current and last velocities for the cell.
  * The u component heads out of the cell to the left, and
  * the v component heads out of the cell to the top.
  * velocities in x, y directions are defined at the center of each face.
  *
  * @author Barry Becker
  */
class CellVelocity {

  /** uip = u(i+0.5, j, k) */
  private val uip = new Array[Double](2)

  /** vjp = v(i, j+0.5, k) */
  private val vjp = new Array[Double](2)

  /** use this to switch between current and last copies of fields. (hack) */
  private var current = 0
  //uip(0) = uip(1)
  //vjp(0) = vjp(1)

  /** global swap of fields (use with care). (hack) */
  def step() {current = 1 - current}

  private[model] def passThrough(): Unit = {
    uip(1 - current) = uip(current) // + dt * forceX;
    vjp(1 - current) = vjp(current) // + dt * forceY;
  }

  private[model] def initializeU(u: Double): Unit = {
    uip(0) = u
    uip(1) = u
  }

  private[model] def initializeV(v: Double): Unit = {
    vjp(0) = v
    vjp(1) = v
  }

  def getU: Double = uip(current)
  def getV: Double = vjp(current)

  private[model] def setCurrentU(u: Double): Unit = {uip(current) = u}

  private[model] def setCurrentV(v: Double): Unit = {vjp(current) = v}
  private[model] def setNewU(newu: Double): Unit = {uip(1 - current) = newu}
  private[model] def setNewV(newv: Double): Unit = {vjp(1 - current) = newv}
  private[model] def incrementU(inc: Double): Unit = {uip(current) += inc}
  private[model] def incrementV(inc: Double): Unit = {vjp(current) += inc}

  def initialize(u: Double, v: Double): Unit = {
    uip(0) = u
    uip(1) = u
    vjp(0) = v
    vjp(1) = v
  }

  override def toString: String = {
    val sb = new StringBuilder("velocity=(")
    sb.append(uip(current)).append(", ").append(vjp(current)).append(")")
    sb.toString
  }
}
